cdap-security/src/test/java/io/cdap/cdap/security/tools/AccessTokenGeneratorTest.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.security.tools;
+
+import io.cdap.cdap.common.http.DefaultHttpRequestConfig;
+import io.cdap.cdap.common.io.Codec;
+import io.cdap.cdap.security.auth.AccessToken;
+import io.cdap.cdap.security.auth.AccessTokenValidator;
+import io.cdap.cdap.security.auth.TokenManager;
+import io.cdap.cdap.security.auth.TokenState;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpRequest;
+import io.cdap.common.http.HttpRequests;
+import io.cdap.common.http.HttpResponse;
+import io.cdap.http.NettyHttpService;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Test for access token generator http endpoint
+ */
+public class AccessTokenGeneratorTest {
+  private static final CompletableFuture<?> future = new CompletableFuture<>();
+  private static final AccessTokenGenerator tokenGenerator = new AccessTokenGenerator();
+  private static NettyHttpService httpService;
+  private static TokenManager tokenManger;
+  private static Codec<AccessToken> tokenCodec;
+
+  @BeforeClass
+  public static void setup() throws Exception {
+    tokenGenerator.init(AccessTokenGenerator.ADDRESS_DEFAULT,
+                                            AccessTokenGenerator.PORT_DEFAULT,
+                                            AccessTokenGenerator.WORKER_POOL_SIZE_DEFAULT,
+                                            AccessTokenGenerator.EXEC_POOL_SIZE_DEFAULT,
+                                            future);
+    httpService = tokenGenerator.getService();
+    httpService.start();
+    tokenManger = tokenGenerator.getTokenManager();
+    tokenCodec = tokenGenerator.getTokenCodec();
+  }
+
+  @AfterClass
+  public static void teardown() throws Exception {
+    httpService.stop();
+    Assert.assertTrue(future.isDone());
+  }
+
+  @Test
+  public void testTokenGenerator() throws Exception {
+    String spec = String.format("http://%s:%s/v3Internal/accesstoken?username=root",
+                                AccessTokenGenerator.ADDRESS_DEFAULT,
+                                AccessTokenGenerator.PORT_DEFAULT);
```

Comment:
Don't use the default port for test. This would make the test flicky. Always use ephemeral port. If you pass `0` as port to the NettyHttpService, it will bind to a random port. All you need is to have the token generator service to provide a visible for test method to get the InetSocketAddress returned by the underlying netty http service.

Reply:
fixed